sky/clouds/service_catalog/common.py

```diff
@@ -111,8 +117,39 @@ def _get_modified_catalogs() -> List[str]:
     return modified_catalog_path_map
 
 
+class LazyDataFrame:
+    """A lazy data frame that reads the catalog on demand."""
+
+    def __init__(self, filename: str):
+        self._filename = filename
+        self._df = None
+
+    def _load_df(self) -> 'pd.DataFrame':
+        if self._df is None:
+            try:
+                self._df = pd.read_csv(self._filename)
+            except Exception as e:  # pylint: disable=broad-except
+                # As users can manually modify the catalog, read_csv can fail.
+                logger.error(f'Failed to read {self._filename}. '
+                             'To fix: delete the csv file and try again.')
+                with ux_utils.print_exception_no_traceback():
+                    raise e
+        return self._df
+
+    def __getattr__(self, name: str):
+        return getattr(self._load_df(), name)
+
+    def __getitem__(self, key):
+        # Delegate the indexing operation to the underlying DataFrame
+        return self._load_df()[key]
+
+    def __setitem__(self, key, value):
+        # Delegate the set operation to the underlying DataFrame
+        self._load_df()[key] = value
+
+
 def read_catalog(filename: str,
-                 pull_frequency_hours: Optional[int] = None) -> pd.DataFrame:
+                 pull_frequency_hours: Optional[int] = None) -> LazyDataFrame:
```

Comment:
Seems like many usage of this result pass it around to downstream callers that take `pd.DataFrame`. Just curious why type checking didn't detect that issue. Example: https://github.com/skypilot-org/skypilot/blob/7847d7e30acc2b83cece8d41c226097221789357/sky/clouds/service_catalog/aws_catalog.py#L170

 Maybe we should make LazyDataFrame inherit `pd.DataFrame` to avoid such confusion?

Reply:
Inheriting from `pd.DataFrame` may not be a good practice for this class, as that may involve many unnecessary variable and methods in this `LazyDataFrame` class, although we only use `self._df` in it.

> Seems like many usage of this result pass it around to downstream callers that take pd.DataFrame. Just curious why type checking didn't detect that issue.

Good point! Not sure if `mypy` has an issue detecting such class mismatch. I changed the caller to use `LazyDataFrame` instead for now.